Add Midnight.js v4.1.1 release notes (#1000)
midnightntwrk/midnight-docs: docs/relnotes/midnight-js/midnight-js-4-1-1.mdx

---
SPDX-License-Identifier: Apache-2.0
copyright: This file is part of midnight-docs. Copyright (C) Midnight Foundation. Licensed under the Apache License, Version 2.0 (the "License"); You may not use this file except in compliance with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0 Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
title: Midnight.js v4.1.1 release notes

> description: Midnight.js is a client library designed to facilitate the development of decentralized applications on the Midnight blockchain.

Author: **Warning — Description (Rule 15)**: The description reads like a generic product summary rather than release-specific SEO copy. It should be 120–160 characters and describe what this release covers.

```suggestion
description: "Release notes for Midnight.js v4.1.1, covering breaking error handling changes, security hardening, and qanet support in testkit-js."
```

displayed_sidebar: sidebar
---

* **Version**: v4.1.1
* **Date**: June 2, 2026
* **Node.js requirement**: >=22

---

## High-level summary

> This release introduces breaking changes to error handling in the indexer public data provider, hardens security for key management and provider URL validation, and adds qanet support in testkit-js. Internal refactoring improves contract governance code organization and type safety.

Author: **Warning — Readability (Rule 44)**: This sentence is 30 words. Consider splitting for clarity.

```suggestion
This release introduces breaking changes to error handling in the indexer public data provider and hardens security for key management and provider URL validation. It also adds qanet support in testkit-js and improves contract governance code organization and type safety.
```

## Audience

This release note is most relevant for developers who:

- Use `IndexerFormattedError.cause` in error handling code (renamed to `.errors`).
- Export or import signing keys through the Midnight.js key management APIs.
- Configure provider URLs for non-loopback environments.
- Run E2E tests against qanet using testkit-js.

## Summary of updates

- Renamed `IndexerFormattedError.cause` to `.errors` for ES2022 `Error.cause` compatibility.
- Applied full password policy to signing key and private state export operations.
- Emitted contract state for `blockHeight` and `blockHash` configurations.
- Hardened error handling in `indexer-public-data-provider` with structured error types.
- Added signing key value validation on import.
- Added warnings for plain HTTP/WS on non-loopback provider URLs.
- Added qanet support via NIGHT/DUST faucet flow in testkit-js.

---

## New features

Below is a detailed breakdown of the new capabilities introduced in this release.

> ### Qanet support in testkit-js

Author: **Warning — Heading sentence case (Rule 31)**: Vale flagged this. "Qanet" should be lowercase as a network name mid-heading, or if it is a proper name, keep it. "testkit-js" is fine as a code identifier.

```suggestion
### qanet support in testkit-js
```

> Testkit-js now supports qanet environments through a NIGHT/DUST faucet flow, enabling E2E testing against the qanet network. The `waitForFunds` address is derived directly from the keystore.

Author: **Suggestion — Passive voice (Rule 8)**: "is derived" should use active voice.

```suggestion
Testkit-js now supports qanet environments through a NIGHT/DUST faucet flow, enabling E2E testing against the qanet network. The keystore derives the `waitForFunds` address directly.
```

### Parameterized compose image versions

Testkit-js compose configurations now accept parameterized image versions, enabling a nightly E2E matrix that tests against multiple component versions.

---

## Breaking changes

### `IndexerFormattedError.cause` renamed to `.errors`

> The ES2022 `Error.cause` slot is contractually a single underlying error, not a collection. Shadowing it with an array of GraphQL errors broke Node's `util.inspect` causal chain, Sentry grouping, and structured loggers. The property is renamed to `.errors` to avoid the conflict.

Author: **Warning — Heading sentence case (Rule 31)**: Vale flagged "renamed to" in the heading.

Also line 58: "is renamed" is passive voice. Rewrite to active.

```suggestion
### Rename `IndexerFormattedError.cause` to `.errors`

The ES2022 `Error.cause` slot is contractually a single underlying error, not a collection. Shadowing it with an array of GraphQL errors broke Node's `util.inspect` causal chain, Sentry grouping, and structured loggers. This release renames the property to `.errors` to avoid the conflict.
```

**Migration**: Replace all references to `IndexerFormattedError.cause` with `.errors`.

### Extracted `correlateDeployTxId` helper

> The contract-action / identifier correlation logic is extracted into a pure helper that takes narrow structural types. `toFinalizedDeployTxData` is now file-local and no longer exported. Use `correlateDeployTxId` instead.

Author: **Suggestion — Passive voice (Rule 8)**: "is extracted" should use active voice.

```suggestion
This release extracts the contract-action / identifier correlation logic into a pure helper that takes narrow structural types. `toFinalizedDeployTxData` is now file-local and no longer exported. Use `correlateDeployTxId` instead.
```

### Non-null assertions replaced with type predicates

Rx filter chains in contract action handling now use type predicates instead of non-null assertions, which may affect code that relied on the previous type narrowing behavior.

---

## Bug fixes and quality improvements

> The following bug fixes and quality improvements are included in this release.

Author: **Suggestion — Passive voice (Rule 8)**: "are included" should use active voice.

```suggestion
This release includes the following bug fixes and quality improvements.
```

### Full password policy on export operations

Export operations for signing keys and private states now enforce the same password policy as other key management operations, closing a security gap.

### Contract state emission for block configurations

> Contract state is now emitted for `blockHeight` and `blockHash` configurations, resolving cases where state updates were missed.

Author: **Suggestion — Passive voice (Rule 8)**: "were missed" should use active voice.

```suggestion
Contract state is now emitted for `blockHeight` and `blockHash` configurations, resolving cases where the system missed state updates.
```

### Hardened indexer error handling

The `indexer-public-data-provider` now uses structured error types with proper error categorization, improving debuggability for GraphQL query failures.

### Signing key validation on import

Imported signing key values are now validated before storage, preventing invalid keys from causing downstream failures.

> ### Plain HTTP/WS warning for non-loopback URLs

Author: **Warning — Heading sentence case (Rule 31)**: Vale flagged "Plain HTTP/WS warning for non-loopback URLs". Should be sentence case.

```suggestion
### Plain HTTP/WS warning for non-loopback URLs
```

(This one is actually already sentence case since HTTP/WS and URLs are abbreviations. The Vale rule is a false positive here, so no change needed.)

Midnight.js now warns when provider URLs use plain HTTP or WebSocket protocols for non-loopback addresses, helping developers identify insecure configurations.

---

## Known issues

> No critical known issues were identified at release time.

Author: **Suggestion — Passive voice (Rule 8)**: "were identified" should use active voice.

```suggestion
No critical known issues exist at release time.
```
